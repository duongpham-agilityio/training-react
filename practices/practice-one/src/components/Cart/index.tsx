import {
  MouseEvent,
  memo,
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState,
} from 'react';
import { useCart, useFetch } from 'hooks';
import { FaTimes } from 'react-icons/fa';
import { useParams } from 'react-router-dom';

// Components
import { ErrorFallback, Heading, Loading } from 'components';
import { Button } from 'components/commons';

// Styles
import styles from './cart.module.css';
import { DEFAULT_DURATION, END_POINT } from '@constants';
import { withErrorBoundary } from 'react-error-boundary';
import { Product } from 'types';

type CartProps = {
  isShowCart: boolean;
  // eslint-disable-next-line no-unused-vars
  onHideCart: (isShowCart: boolean) => void;
};

const Cart = (props: CartProps) => {
  const { isShowCart, onHideCart } = props;
  const { id } = useParams();
  const {
    data: product,
    isLoading,
    error,
    mutate,
  } = useFetch<Product>(id ? `${END_POINT.PRODUCTS}/${id}` : null);
  const { getProductFromCart, removeAll, removeOne, payment } = useCart();
  const cartRef = useRef<HTMLDivElement>(null);
  const [data, setData] = useState(getProductFromCart() || []);

  //---------------- Calculating data ----------------

  const total = useMemo(() => {
    const total = data.reduce((prev, current) => {
      return prev + current.price * current.quantity;
    }, 0);

    return total;
  }, [data]);

  //---------------- Handlers ----------------

  /**
   * Calculate total amount
   */
  const bindCloseCart = useCallback(() => {
    const cartElement = cartRef.current;

    cartElement?.classList.remove(styles.open);
    cartElement?.classList.add(styles.close);

    setTimeout(() => onHideCart(false), DEFAULT_DURATION);
  }, [onHideCart]);

  /**
   * Remove one product in cart
   * @param id id of product
   */
  const handleRemoveOne = (id: number) => {
    const carts = data.filter((cart) => cart.id !== id);

    setData(carts);
    removeOne(id);
  };

  /**
   *  Clear all products in cart
   */
  const clearCart = () => {
    setData([]);
    removeAll();
  };

  const handlePayment = async () => {
    setData([]);
    if (product) {
      payment(product.id, (productWithStore) => {
        const quantityCurrentProduct = productWithStore?.quantity || 0;
        const newData = {
          ...product,
          quantity: product.quantity - quantityCurrentProduct,
        };

        mutate(newData, {
          populateCache: (newData) => newData,
          revalidate: false,
        });
      });
    }
  };

  useEffect(() => {
    document.addEventListener('click', bindCloseCart);

    return () => document.removeEventListener('click', bindCloseCart);
  }, [bindCloseCart]);

  if (isLoading) return <Loading />;

  if (error) {
    return <ErrorFallback error={error} resetErrorBoundary={() => {}} />;
  }

  return (
    <>
      {/* add class 'open' or 'close */}
      <section
        ref={cartRef}
        className={`${styles.cartWrapper} ${
          isShowCart ? styles.open : styles.close
        }`}
        onClick={(e: MouseEvent) => e.stopPropagation()}
      >
        <div className={styles.cartHeder}>
          <Heading>
            <span className={styles.cartTitle}>My Basket</span>
            <span className={styles.cartQuantity}>{`( ${data?.length} item${
              // eslint-disable-next-line no-constant-condition
              data && data.length > 1 ? 's' : ''
            })`}</span>
          </Heading>

          <div className={styles.cardHeaderActions}>
            <Button
              description="Close"
              variant="primary"
              size="sm"
              className={styles.btn}
              onClick={bindCloseCart}
            />
            <Button
              description="Clear Basket"
              size="sm"
              className={styles.btn}
              onClick={clearCart}
              variant={data.length ? 'primary' : 'default'}
              disabled={!data.length}
            />
          </div>
        </div>

        <div className={styles.cartBody}>
          {data.map(({ id, imageURL, title, quantity, price }) => {
            return (
              <div className={styles.cartItem} key={id}>
                <img
                  src={`/${imageURL}`}
                  alt={title}
                  className={styles.cartImage}
                />
                <div className={styles.cartInformation}>
                  <Heading description={title} className={styles.productName} />
                  <div className={styles.cartDetails}>
                    <div className={styles.detailItem}>
                      <Heading
                        tag="h3"
                        description="Quantity"
                        className={styles.detailHeading}
                      />
                      <p className={styles.detailValue}>{quantity}</p>
                    </div>
                    <div className={styles.detailItem}>
                      <Heading
                        tag="h3"
                        description="Size"
                        className={styles.detailHeading}
                      />
                      <p className={styles.detailValue}>28 mm</p>
                    </div>
                  </div>
                </div>
                <p className={styles.productPrice}>${price}</p>
                <div className={styles.cartItemActions}>
                  <Button
                    leftIcon={<FaTimes />}
                    size="sm"
                    className={styles.btn}
                    onClick={() => handleRemoveOne(id)}
                  />
                </div>
              </div>
            );
          })}
        </div>

        <div className={styles.cartFooter}>
          <div className={styles.cartTotal}>
            <p className={styles.totalText}>Subtotal Amout:</p>
            <p className={styles.total}>${total}</p>
          </div>
          <Button
            description="check out"
            className={styles.btn}
            size="lg"
            variant={data.length ? 'primary' : 'default'}
            disabled={!data.length}
            onClick={handlePayment}
          />
        </div>
      </section>
    </>
  );
};

export default withErrorBoundary(memo(Cart), {
  FallbackComponent: ErrorFallback,
});
