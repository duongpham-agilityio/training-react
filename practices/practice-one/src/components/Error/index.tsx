import { FallbackProps } from 'react-error-boundary';

// Icons
import { AiOutlineWarning } from 'react-icons/ai';

// Styles
import styles from './error.module.css';
import container from 'styles/commons/container.module.css';

type ErrorProps = FallbackProps;

const ErrorFallback = ({ error }: ErrorProps) => {
  return (
    <div className={`${styles.error} ${container.container}`}>
      <AiOutlineWarning className={styles.icon} />
      <p className={styles.message}>{error.message}</p>
    </div>
  );
};

export default ErrorFallback;
