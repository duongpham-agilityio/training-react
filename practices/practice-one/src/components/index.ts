import { lazy } from 'react';

const ErrorFallback = lazy(() => import('./Error'));
const Heading = lazy(() => import('./Heading'));
const Toast = lazy(() => import('./Toast'));
const Card = lazy(() => import('./Card'));
const Cart = lazy(() => import('./Cart'));
const Filter = lazy(() => import('./Filter'));
const Grid = lazy(() => import('./Grid'));
const Loading = lazy(() => import('./Loading'));
const Select = lazy(() => import('./Select'));

export {
  ErrorFallback,
  Heading,
  Toast,
  Card,
  Cart,
  Filter,
  Grid,
  Loading,
  Select,
};
