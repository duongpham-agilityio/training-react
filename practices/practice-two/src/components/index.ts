export {
  default as Notification,
  type NotificationProps,
} from './Notification';

export { default as Popup, type PopupProps } from './Popup';
export { default as Card, type CardProps } from './Card';
export { default as CardSkeleton } from './Card/Skeleton';
export { default as Error } from './Error';
