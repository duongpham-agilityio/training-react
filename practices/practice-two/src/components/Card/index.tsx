import { MouseEvent, memo } from 'react';
import { Link, LinkProps } from 'react-router-dom';

// Component
import { Button, Heading } from 'components/commons';

// Styles
import styles from 'components/Card/index.module.css';

// Assets
import NextIcon from 'assets/icons/next.svg';
import TrashIcon from 'assets/icons/trash.svg';

export type CardProps = Omit<LinkProps, 'to'> & {
  href?: string;
  title: string;
  description: string;
  publishedDate: string;
  imageUrl: string;
  onDeleteBook?: (_event: MouseEvent) => void;
};

const Card = (props: CardProps) => {
  const {
    title,
    description,
    publishedDate,
    imageUrl,
    href = '',
    onDeleteBook,
  } = props;

  return (
    <Link to={href} className={styles.card}>
      <div className={styles.cardItem}>
        <Heading label={title} className={styles.heading} />
        <p className={styles.description}>{description}</p>
        <div className={styles.action}>
          <Button leftIcon={NextIcon} variant="primary" />
          <Button
            leftIcon={TrashIcon}
            variant="danger"
            onClick={onDeleteBook}
          />
        </div>
      </div>
      <div className={styles.cardItem}>
        <p className={styles.time}>{publishedDate}</p>
        {imageUrl ? (
          <img src={imageUrl} alt={title} className={styles.image} />
        ) : (
          <div className={`${styles.image} ${styles.fake}`}></div>
        )}
      </div>
    </Link>
  );
};

export default memo(Card);
