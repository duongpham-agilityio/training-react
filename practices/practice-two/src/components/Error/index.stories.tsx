import { Meta, Story } from '@storybook/react';

// Component
import { Error } from 'components';

export const Template: Story = () => <Error />;

export default {
  title: 'Practice/components/Error',
  component: Error,
} as Meta;
