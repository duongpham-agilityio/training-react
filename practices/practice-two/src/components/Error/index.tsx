// Component
import { Heading } from 'components/commons';

// Constants
import { MESSAGES } from '@constants';

// Styles
import styles from 'components/Error/index.module.css';

export interface ErrorProps {
  title?: string;
  description?: string;
}

const Error = ({
  title = MESSAGES.ERROR_TITLE,
  description = MESSAGES.ERROR_DESCRIPTION,
}: ErrorProps) => (
  <section className={styles.empty}>
    <Heading label={title} className={styles.emptyTitle} size="sm" />
    <p className={styles.emptyDescription}>{description}</p>
  </section>
);

export default Error;
