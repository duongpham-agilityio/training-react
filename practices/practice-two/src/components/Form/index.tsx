import { MouseEvent, memo, useCallback } from 'react';
import isEqual from 'react-fast-compare';

// Hooks
import { useBookForm, FormData } from 'hooks';

// Components
import { Button, Heading, Input } from 'components/commons';
import BookRecommend from './Recommend';
import ChooseImage from './ChooseImage';

// Styles
import styles from 'components/Form/index.module.css';

export type FormProps = {
  value: FormData;
  title?: string;
  className?: string;
  type?: 'update' | 'create';
  onClose?: () => void;
  onSubmit: (data: FormData) => Promise<void>;
};

const Form = (props: FormProps) => {
  const {
    value: data,
    title = 'Create a new book marker',
    className = '',
    type = 'create',
    onClose,
    onSubmit,
  } = props;
  const {
    value: { author, description, imageURL, name, publishDate, imageName },
    refImage,
    isUpload,
    isSubmit,
    booksRecommended,
    handleSelectRecommended,
    resetRecommended,
    onChange,
    onSubmit: onSubmitOnHook,
  } = useBookForm(data, type, onSubmit);

  const chooseImage = useCallback(() => {
    refImage.current?.click();
  }, [refImage]);

  const resetRecommendedData = useCallback(
    (e: MouseEvent) => {
      e.stopPropagation();

      resetRecommended([]);
    },
    [resetRecommended]
  );

  return (
    <section className={styles.overlay} onClick={onClose}>
      <form
        className={`${styles.form} ${className}`}
        action="#"
        method="POST"
        onSubmit={onSubmitOnHook}
        onClick={resetRecommendedData}
        data-testid="book-form"
      >
        <Heading label={title} className={styles.heading} />

        <section className={styles.content}>
          <div className={styles.formItem}>
            <div className={styles.formField}>
              <label className={styles.label} htmlFor="bookName">
                Book name
              </label>
              <Input
                value={name}
                placeholder="Book name"
                name="name"
                id="bookName"
                onChange={onChange}
              />

              {!!booksRecommended.length && (
                <BookRecommend
                  books={booksRecommended}
                  onSelect={handleSelectRecommended}
                />
              )}
            </div>
            <div className={styles.formField}>
              <label className={styles.label} htmlFor="author">
                Author
              </label>
              <Input
                value={author}
                placeholder="Author"
                name="author"
                onChange={onChange}
              />
            </div>
          </div>

          <div className={styles.formItem}>
            <div className={styles.formField}>
              <label className={styles.label} htmlFor="publishedDate">
                Published date
              </label>
              <Input
                type="date"
                value={publishDate}
                name="publishDate"
                onChange={onChange}
              />
            </div>
            <div className={`${styles.formField} ${styles.upload}`}>
              <div className={styles.uploadItem}>
                <label className={styles.label} htmlFor="imageURL">
                  Upload
                </label>
                <Input
                  ref={refImage}
                  hidden
                  className={styles.input}
                  type="file"
                  name="imageURL"
                  id="imageURL"
                  accept=".png,.jpeg"
                  onChange={onChange}
                />

                <ChooseImage name={imageName} chooseHandler={chooseImage} />
              </div>
              <div className={styles.uploadItem}>
                {imageURL && (
                  <img
                    src={imageURL}
                    alt="image book"
                    className={styles.image}
                  />
                )}
              </div>
            </div>
          </div>

          <div className={styles.formItem}>
            <div className={styles.formField}>
              <label className={styles.label} htmlFor="bookName">
                Description
              </label>
              <Input
                variant="area"
                className={styles.text}
                value={description}
                placeholder="Book name"
                name="description"
                onChange={onChange}
              />
            </div>
          </div>

          <div className={styles.action}>
            <Button
              label="Cancel"
              width="w-lg"
              border="b-lg"
              onClick={onClose}
              disabled={isSubmit}
            />
            <Button
              disabled={isUpload || isSubmit}
              type="submit"
              label="Save"
              variant={isUpload ? 'default' : 'primary'}
              width="w-lg"
              border="b-lg"
            />
          </div>
        </section>
      </form>
    </section>
  );
};

export default memo(Form, isEqual);
