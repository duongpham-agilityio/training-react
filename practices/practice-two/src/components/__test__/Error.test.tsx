import { render } from '@testing-library/react';

// Mocks
import './mocks/button.test';
import './mocks/image.test';

// Components
import { Error } from 'components';

// Types
import { ErrorProps } from 'components/Error';

const setup = (props: ErrorProps) => render(<Error {...props} />);

describe('Error component', () => {
  it('Match snapshot', () => {
    const { container } = setup({});

    expect(container).toMatchSnapshot();
  });

  it('Render with props', () => {
    const mockProps: ErrorProps = {
      title: 'Error',
      description: 'Error description',
    };
    const { getByText } = setup(mockProps);

    expect(getByText(mockProps.title ?? '')).toBeInTheDocument;
    expect(getByText(mockProps.description ?? '')).toBeInTheDocument;
  });
});
