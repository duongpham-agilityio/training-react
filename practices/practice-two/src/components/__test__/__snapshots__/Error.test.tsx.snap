// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error component Match snapshot 1`] = `
<div>
  <section>
    <h2
      class="undefined undefined "
    >
      Oops! Something went wrong
    </h2>
    <p>
      Please, help me reload pages.
    </p>
  </section>
</div>
`;
